### ISSUE TITLE
Add ITestCase for Trial tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2898]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2898|
|Type|enhancement|
|Created|2007-11-13 22:10:04Z|
|Branch|https://github.com/twisted/twisted/tree/itestcase-2898-2|

Define an interface for tests as used by the test framework and then do pyunit compat with an adapter.

<details><summary>Searchable metadata</summary>

```
trac-id__2898 2898
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_itestcase_2898_2 branches/itestcase-2898-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1194991804000000 1194991804000000
changetime__1195242065000000 1195242065000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [21751]) Branching to 'itestcase-2898'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to @jml|
|-|-|

Doesn't work on Python 2.3.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

Fixed in Python 2.3.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @jml|
|-|-|

Global comments:
 * the new warnings are not tested. The test suite now prints warnings (but not only from these changes, unfortunately). At least, the new warnings emitted should be filtered. Also, the text of the warnings are incomplete, it should explicitly say what is deprecated, and if there is an alternative, point to it.
 * test_doctest.py and test_loader.py seem to contain unrelated changes

In test_pyunitcompat.py:
 * the docstring of test_id is not clear

In runner.py:
 * Why did you remove completely the docstring of ```DocTestSuite```? The fact it's deprecated doesn't mean it should be undocumented :)
 * I don't understand the change in isTestCase. Isn't the goal to use ```ITestCase``` more?

In unittest.py:
 * ```_AdaptedReporter``` and its methods have no docstring

In itrial.py:
 * Why setUp and tearDown have been removed?
 * it's unfortunate that ```__call__``` is introduced, but that may depends on stdlib unittest.

Also, if you can explain why after deprecating it, we now reuse it, that would be great. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

(In [21801]) Remove unnecessary test. Refs [#2898](https://github.com/twisted/twisted/issues/2898).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml commented|
|-|-|

(In [21802]) Branching to 'itestcase-2898-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml commented|
|-|-|

I'll add a more general note before I continue fixing the details.

The goal of this branch is to provide an expected interface for unit tests, and to use this interface to handle incompatibilities. By coincidence, there is a deprecated, stupid interface with the name that we want, that is only used as a way of marking classes that we want to try to load as tests.

This interface must be the public methods and attributes that are used by _testing frameworks_. In other words, given a test case object, what can I do with it?

For this reason `setUp` and `tearDown` are not included. They are part of a separate contract provided by the `run()` method of many `TestCase`-style classes.

The interface cannot just define the contract that Trial's test framework relies on, it must also define the contract that other test frameworks can be expected to rely on. For our purposes, "other test frameworks" means Python standard library `unittest`. Trial tests should be useable in stdlib test suites: that's why we must implement `__call__`.

So we have an interface that defines exactly what Trial expects from a test case (`ITestCase`), and we want to support tests that don't yet implement that interface (`pyunit.TestCase`). We want to do this using an adapter.

That means we can't use `ITestCase` to _find_ the pyunit test cases. Checking for subclasses of `pyunit.TestCase` is the simplest thing that is correct.

This is a long explanation. Its complexity is one of the reasons that I'm breaking up my work on introducing test decorators.

I hope this helps.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml set owner to @therve|
|-|-|

 * The unrelated changes are there from when I've been experimenting with those files and fixing up the code to be standards compliant. Those changes are staying there.
 * Restored docstring to `DocTestSuite`.
 * Basically, no one should be using any of the stuff in `twisted.trial.runner`. It's support code for the testing framework. `DocTestSuite` was just a technique for Trial to support doctests. Now we are switching to a different, more general technique that doesn't have an analog for `DocTestSuite`.
 * Docstrings added for `_AdaptedReporter`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

Well, some of my previous comments are still true: new warnings are being printed, whereas they should be suppressed, and there is no test for warnings. If you don't care about that, fine, but I'm not able to review that code then.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

The tests printing new warnings are ```twisted.trial.test.test_doctest.TestRunners.test_correctCount``` and ```twisted.trial.test.test_loader.LoaderTest.test_loadDifferentNames```. Although, they may appear somewhere else after these ones are corrected (I think? warnings is so strange).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml commented|
|-|-|

Warnings fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @jml|
|-|-|

Some questions:

  * Is ```iterateTests``` intended to be public?
  * Why does ```_AdaptedReporter``` only implement ```addError``` and not the other methods which take a test case?
  * ```suiteVisit``` and ```DocTestSuite``` in ```runner.py``` changed to adapt objects by calling ```ITestCase``` instead of explicitly instantiating the right adapter class.  Is there a deep reason for this, or is it only to avoid duplicating the logic which associates ```PyUnitTestCase``` with ```pyunit.TestCase``` which is already present in ```unittest.py``` (in the ```registerAdapter``` calls)?  If so, cool.  Generally it's nicer to explicitly instantiate an adapter class in code which definitely knows what kind of thing is being adapted, though.

One simple thing:

  * The docstring for ```ITestCase.failureException``` might want to say _exception class_ instead of _exception_

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_17"></a>|@jml commented|
|-|-|

In order:
 * Not really. Renamed to `_iterateTests`
 * It's all that's needed to make it work. There'll be a branch soon that provides a full implementation.
 * `_PyUnitTestAdapter` is different from `PyUniTestCase`. In these cases it's not actually known which classes are being adapted.
 * Docstring updated.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_18"></a>|@jml set status to `closed`|
|-|-|

(In [21827]) Define `ITestCase` to be the interface that test frameworks expect of a test.

 * Author: jml
 * Reviewers: therve, exarkun
 * Fixes [#2898](https://github.com/twisted/twisted/issues/2898)

`ITestCase` is re-purposed to define what you can expect from a test object.
We use this interface to adapt pyunit tests to things usable in Trial.

In the process we also clarify what's needed for compatibility with doctest
and pyunit.

This branch also fixes [#1950](https://github.com/twisted/twisted/issues/1950).

Refs [#2739](https://github.com/twisted/twisted/issues/2739).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
